File: src/structural/facade/python/classes/facade/builder_facade.py
import sys

sys.path.insert(0, './src/creational/builder/python')


if True:
    from classes.builder.main_dish_builder import MainDishBuilder
    from classes.builder.vegan_dish_builder import VeganDishBuilder
    from classes.meal_box import MealBox
    from classes.meals import Beans, Meat, Rice


class BuilderFacade():

    def __init__(self):
        self.main_dish_builder = MainDishBuilder()
        self.vegan_dish_builder = VeganDishBuilder()

        self.RICE = Rice('Arroz', 5)
        self.BEANS = Beans('Feijão', 10)
        self.MEAT = Meat('Carne', 20)

    def make_meal_with_two_builder(self):
        meal1 = self.main_dish_builder.make_meal().make_dessert().get_meal()
        print(meal1)
        print(meal1.get_price())

        meal2 = self.vegan_dish_builder.make_meal().get_meal()
        print(meal2)
        print(meal2.get_price())

    def make_meal_without_builder(self):

        meal_box = MealBox()
        meal_box.add(self.RICE, self.BEANS, self.MEAT)

        print(meal_box._children)
        print(meal_box.get_price())

File: src/structural/facade/typescript/classes/facade/BuilderFacade.ts
import { MainDishBuilder } from '../../../../../creational/builder/typescript/classes/builders/MainDishBuilder';
import { MealBox } from '../../../../../creational/builder/typescript/classes/MealBox';
import {
  Beans,
  Meat,
  Rice,
} from '../../../../../creational/builder/typescript/classes/Meals';
import { BuilderFacadeProtocol } from '../interfaces/BuilderFacadeProtocol';

export class BuilderFacade implements BuilderFacadeProtocol {
  readonly mainDishBuilder = new MainDishBuilder();

  readonly rice = new Rice('Arroz', 5);
  readonly beans = new Beans('Feijão', 10);
  readonly meat = new Meat('Carne', 20);

  makeMealWithTwoBuilders(): void {
    const meal1 = this.mainDishBuilder.makeMeal().makeDessert().getMeal();
    console.log(meal1);
    console.log(meal1.getPrice());

    const meal2 = this.mainDishBuilder.makeMeal().makeDessert().getMeal();
    console.log(meal2);
    console.log(meal2.getPrice());
  }

  makeMealWithoutBuilder(): void {
    const mealBox = new MealBox();
    mealBox.add(this.rice, this.beans, this.meat);
    console.log(mealBox);
    console.log(mealBox.getPrice());
  }
}
